torch/csrc/distributed/c10d/init.cpp

```diff
@@ -975,18 +976,21 @@ that adds a prefix to each key inserted to the store.
 
           .def(
               "broadcast",
-              &::c10d::ProcessGroup::broadcast,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
+                 const std::vector<at::Tensor>& tensors,
+                 ::c10d::BroadcastOptions opts) {
+                return ::c10d::ops::broadcast(self, tensors, opts);
```

Comment:
This does capture all the broadcast calls from the python side, however I think the ideal place is to call the ops::broadcast directly here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1182. I'm worried that the traced version of the program won't have the python validation? For example, there is this logic on the python side that is important:

```
    if _rank_not_in_group(group):
        _warn_not_in_group("broadcast")
        return
```

I guess the python side of things would still run with something like TorchDynamo, but if there is something that traces a program in a language agnostic IR and would like to run the IR in C++, we would run into issues.

This can be done in a follow up PR since it is probably a much bigger change, but I feel our python API should directly call the aten op and not have any python logic before that.

Reply:
> I'm worried that the traced version of the program won't have the python validation?

This depends on the tracer. For dynamo, we should ensure that it is replacing the python validation with its own guards.  (That may actually not be the case today as dynamo mostly has special logic for tensors/nn.modules and friends.

If we don't use dynamo, e.g. for a direct AotAutograd we'd need another policy for how to ensure safety, but it wouldn't happen by default.